tensorflow/core/kernels/data/experimental/io_ops.cc

```diff
@@ -262,7 +262,12 @@ class LoadDatasetOp::Dataset : public DatasetBase {
     explicit Iterator(const Params& params)
         : DatasetIterator<Dataset>(params) {}
 
-    ~Iterator() override { input_->Unref(); }
+    ~Iterator() override {
+      // Ensure not cleaning up uninitialized input_
+      if(input_) {
+        input_->Unref();
```

Comment:
Thanks for the change. Do you know why we need this? I don't see this class calls `Ref()`. There is a `Unref` call in the base iterator.
https://github.com/tensorflow/tensorflow/blob/a9cf3a0e4b419630f0183b0cc4e48e0641a62721/tensorflow/core/framework/dataset.cc#L701

Reply:
@yangustc07  thank you for the review.
This is the corresponding call to Ref(): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/experimental/io_ops.cc#L338

Reply:
And this Ref() and the input_ initialization does not happen because of a cancelled prefetch:
```bash
tensorflow/core/kernels/data/experimental/io_ops.cc:282] Status Cancelled:
        For thread id 1397804166 thread name: tf_data_iterator_resource/1397804166*** Begin stack trace ***
        tensorflow::CurrentStackTrace()
        tensorflow::data::experimental::LoadDatasetOp::Dataset::Iterator::Initialize(tensorflow::data::IteratorContext*)
        tensorflow::data::DatasetBase::MakeIterator(tensorflow::data::IteratorContext*, tensorflow::data::IteratorBase const*, std::string const&, std::unique_ptr<tensorflow::data::IteratorBase, std::default_delete<tensorflow::data::IteratorBase> >*) const
        tensorflow::data::ShuffleDatasetOpBase::ShuffleDatasetBase::Iterator::GetNextInternal(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::DatasetBaseIterator::GetNext(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::BatchDatasetOp::Dataset::Iterator::GetNextInternal(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::DatasetBaseIterator::GetNext(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::RepeatDatasetOp::Dataset::ForeverIterator::GetNextInternal(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::DatasetBaseIterator::GetNext(tensorflow::data::IteratorContext*, std::vector<tensorflow::Tensor, std::allocator<tensorflow::Tensor> >*, bool*)
        tensorflow::data::PrefetchDatasetOp::Dataset::Iterator::PrefetchThread(std::shared_ptr<tensorflow::data::IteratorContext> const&)

        std::_Function_handler<void (), std::_Bind<void (*(std::function<void ()>, std::shared_ptr<tensorflow::Notification>))(std::function<void ()> const&, std::shared_ptr<tensorflow::Notification>)> >::_M_invoke(std::_Any_data const&)
        tensorflow::UnboundedWorkQueue::PooledThreadFunc()
```

Reply:
I see, it seems this `Ref` is not needed for the same reason mentioned in https://github.com/tensorflow/tensorflow/commit/858a5698a7b4ee95befa2e9c3d7aaa0a8170ec54. What do you think?

Reply:
Actually removing Ref results in a segfault without a stracktrace in this case. I'm still investigating that.

Reply:
@yangustc07 Did you have any other feedback?

Reply:
I'm trying this test on my machine. I can produce the same segfault. I'll update you soon. I'm sorry for the inconvenience.

Reply:
Yes, I agree with you. `Ref()` is needed here because `reader_output` is destroyed in `InitializeInput`.

Thanks for the fix!